#!/bin/bash
#PBS -P myrtaceae
#PBS -N HMM
#PBS -l select=1:ncpus=2:mem=10GB
#PBS -l walltime=10:00:00
#PBS -e ./hmm_eror.txt
#PBS -o ./hmm_output.txt
#PBS -M peri.tobias@sydney.edu.au
#PBS -m abe

#Parameters
wdir=/scratch/myrtaceae
outdir=$wdir/2.results

#Load modules
module load hmmer/3.3
module load bedtools/2.29.2
module load seqkit/0.10.1
module load clustal-omega/1.2.4
module load python/3.6.5

##Also all additional scripts in home directory or path specified. Scripts required are: exorf_or.py, headers_to_bed.py, make_bed_hmmOut.awk, translate.py. This script will run through all the steps and output an exended nt fasta for NBARC regions in genome (called ${prefix_name}_NBARC_20kb.fasta) plus intermediate files which may/may not be useful. Scripts should be in wdir/0.scripts/ Genome.fasta, NLR-annotator bedfile and HMMs in wdir/1.data/ and all outputs to wdir/2.results.

##We pass the genome and prefix for outputs when submitting the pbs script, for example: qsub -v prefix_name=E_grandis,input_fasta=E_grandis.fasta myrtaceae_hmm.pbs

cd $PBS_O_WORKDIR

##Step A - run HMM - we use the Egrandis TIR and nonTIR HMMs (Christie et al. 2016) to identify NBARC regions within our chosen Myrtaceae genome.
nhmmer $wdir/1.data/EG_nonTIRhmm $wdir/1.data/$input_fasta > $outdir/${prefix_name}_nonTIRout
nhmmer $wdir/1.data/EG_TIRhmm $wdir/1.data/$input_fasta > $outdir/${prefix_name}_TIRout

## Output from first step is used to make a bed file to extract NBARC sequences from the original genome fasta. We then combine the two fasta outputs and remove duplicates.
make_bed_hmmOut.awk $outdir/${prefix_name}_nonTIRout > $outdir/${prefix_name}_nonTIR.bed
make_bed_hmmOut.awk $outdir/${prefix_name}_TIRout > $outdir/${prefix_name}_TIR.bed
bedtools getfasta -s -fi $wdir/1.data/$input_fasta -bed $outdir/${prefix_name}_nonTIR.bed -fo $outdir/${prefix_name}_nonTIR.fasta
bedtools getfasta -s -fi $wdir/1.data/$input_fasta -bed $outdir/${prefix_name}_TIR.bed -fo $outdir/${prefix_name}_TIR.fasta

##Step B - Use the following method to make *species-specific* NBARC hmms from Myrtaceae genome file using the previous output as training set.

## This awk script will extract first 200 (or designated numeral for variable NSEQS) sequences from previously output nonTIR and TIR.fasta. Then we combine these 400 sequences, remove duplicates and align with clustalo to make stockholm format for next steps of hmmer. This use of E.grandis HMM training set for *species-specific* HMM to extract NBS-LRR genes increases the output of NBARCs.
NSEQS=200
awk "/^>/ {n++} n>$NSEQS {exit} 1" $outdir/${prefix_name}_nonTIR.fasta > $outdir/${prefix_name}_nonTIR_200.fasta
awk "/^>/ {n++} n>$NSEQS {exit} 1" $outdir/${prefix_name}_TIR.fasta > $outdir/${prefix_name}_TIR_200.fasta
cat $outdir/${prefix_name}_nonTIR_200.fasta $outdir/${prefix_name}_TIR_200.fasta > $outdir/${prefix_name}_NBARC400.fasta
seqkit rmdup -D duplicates -n $outdir/${prefix_name}_NBARC400.fasta > $outdir/${prefix_name}_NBARC.fasta
clustalo -i  $outdir/${prefix_name}_NBARC.fasta -o  $outdir/${prefix_name}_NBARC.sto --outfmt=st
hmmbuild -nucleic  $wdir/1.data/${prefix_name}.hmm  $outdir/${prefix_name}_NBARC.sto
nhmmer $wdir/1.data/${prefix_name}.hmm $wdir/1.data/$input_fasta > $outdir/${prefix_name}_NBARCout
make_bed_hmmOut.awk $outdir/${prefix_name}_NBARCout > $outdir/${prefix_name}_NBARC.bed
bedtools getfasta -s -fi $wdir/1.data/$input_fasta -bed $outdir/${prefix_name}_NBARC.bed -fo $outdir/${prefix_name}_NBARC_nt.fasta

##Step C - Translate Nucleotide NBARC sequences - 6-frame translations plus longest ORF frame output to fasta for downstream analysis, including extracting extended regions around NBARC domains from original genome and incorporating NLR-annotator defined loci.
translate.py $outdir/${prefix_name}_NBARC_nt.fasta $outdir/${prefix_name}_NBARC_aa.fasta
##Extract headers from NBARC_aa.fasta
grep '^>' $outdir/${prefix_name}_NBARC_aa.fasta > $outdir/${prefix_name}_NBARC_headers.txt
##Use the headers from the amino acid NBARC fasta to make an extended bedfile that incorporates +/- (20,000 nt) up and downstream of NBARC with python script so that the extended nucleotide fasta sequences can be extracted from original genome file. This can be used to identify other NLR type motifs such as TIR domains,LRR etc.
headers_to_bed.py $outdir/${prefix_name}_NBARC_headers.txt 20000 $wdir/1.data/$input_fasta $outdir/${prefix_name}_NBARC_20kbflanking.bed
#bedtools merge requires that you presort your data by chromosome and then by start position, as below.
sort -k1,1 -k2,2n $outdir/${prefix_name}_NBARC_20kbflanking.bed > $outdir/${prefix_name}_NBARC_20kbflanking_sort.bed
#At this point we can incorporate the data output from nlr-annotator. To do this we need to standardise the bed file from nlr-annotator to match column data with ${prefix_name}_NBARC_20kbflanking_sort.bed. We then combine files and merge
gawk 'BEGIN { OFS = "\t"} ; { col5 = ($6 == "+" ? "forward" : "reverse");  print $1, $2, $3, $1, col5, $6 }' $wdir/1.data/${prefix_name}_nlr.bed > $outdir/${prefix_name}_new_nlr.bed
sort -k1,1 -k2,2n $outdir/${prefix_name}_new_nlr.bed > $outdir/${prefix_name}_new_sort_nlr.bed
cat $outdir/${prefix_name}_NBARC_20kbflanking_sort.bed $outdir/${prefix_name}_new_sort_nlr.bed > $outdir/${prefix_name}_NBARC_20kb.bed
bedtools merge -s -d 1 -c 5,4,6 -o count,distinct,last -i $outdir/${prefix_name}_NBARC_20kb.bed > $outdir/${prefix_name}_NBARC_20kb_merge.bed
bedtools getfasta -s -fi $wdir/1.data/$input_fasta -bed $outdir/${prefix_name}_NBARC_20kb_merge.bed -fo $outdir/${prefix_name}_NBARC_20kb.fasta

##Step D - Take the previous output ($outdir/${prefix_name}_NBARC_20kb.fasta)  and annotate with Supplemetary Data https://doi.org/10.1101/2020.07.08.193961 (NLR Refseqs aa) as hints in --epmode. Then remove the * from the amino acid augusutus output and scan with interproscan-5 to identify TIR, LRR coils etc. for a complete list and fasta outptut of NLR putative genes. Interproscan will also identify other important domains within proximity to NLRs.

##Step E - The output from interproscan needs to be correlated with the augustus.gff3 and to the original genome co-ordinates. Scripts for extracting the gene IDs and chromosome locations as gff3 files are: run_NBARC.sh, run_NBLRR.sh

module unload seqkit/0.10.1
module unload hmmer/3.3
module unload bedtools/2.29.2
module unload clustal-omega/1.2.4
module unload python/3.6.5
