#!/bin/bash
#PBS -P myrtaceae
#PBS -N HMM
#PBS -l select=1:ncpus=2:mem=10GB
#PBS -l walltime=10:00:00
#PBS -e ./hmm_eror.txt
#PBS -o ./hmm_output.txt
#PBS -M peri.tobias@sydney.edu.au
#PBS -m abe

#Parameters
wdir=/project/myrtaceae
outdir=$wdir/2.results

#Load modules
#module load hmmer/3.2
module load hmmer/3.3
module load bedtools/2.29.2
module load seqkit/0.10.1
module load clustal-omega/1.2.4
module load python/3.6.5

cd $PBS_O_WORKDIR

##We need the following scripts to be located in home/dir - make_bed_hmmOut.awk,translate.py
##We need the following files to be located in $wdir/1.data/ - EG_nonTIRhmm,EG_TIRhmm,euc_NLR_aa.fasta,E_grandis_NLR_cds.fasta
##We  pass the genome and prefix for outputs when submitting the pbs script, for example: 
# qsub -v prefix_name=E_grandis_NBARC,final_name=eucalyptus_NBARC,input_fasta=E_grandis_NLR.fasta,input_aa.fasta=euc_NLR_aa.fasta NBARC_aa_clade_process.pbs

##Use nhmmer to first extract NBARC regions from E_grandis_NLR_cds.fasta using EG_TIR and nonTIRhmm. Extract regions from cds fasta and then translate for E_grandis_NBARC_domains only aa.fasta file. Use the NBARC_aa.fasta to build aa.hmm to use on the euc_NLR_aa.fasta (all the combined eucalytus NLR_aa.fastas).

##Step A - run HMM - we use the Egrandis TIR and nonTIR HMMs (Christie et al. 2016) to identify NBARC regions within our E_grandis_NBLRR_cds.fasta.
nhmmer $wdir/1.data/EG_nonTIRhmm $wdir/1.data/$input_fasta > $outdir/${prefix_name}_nonTIRout
nhmmer $wdir/1.data/EG_TIRhmm $wdir/1.data/$input_fasta > $outdir/${prefix_name}_TIRout

##Step B - Output from first step is used to make a bed file to extract NBARC sequences from the E_grandis_NBLRR_cds.fasta.
make_bed_hmmOut.awk $outdir/${prefix_name}_nonTIRout > $outdir/${prefix_name}_nonTIR.bed
make_bed_hmmOut.awk $outdir/${prefix_name}_TIRout > $outdir/${prefix_name}_TIR.bed
bedtools getfasta -s -fi $wdir/1.data/$input_fasta -bed $outdir/${prefix_name}_nonTIR.bed -fo $outdir/${prefix_name}_nonTIR.fasta
bedtools getfasta -s -fi $wdir/1.data/$input_fasta -bed $outdir/${prefix_name}_TIR.bed -fo $outdir/${prefix_name}_TIR.fasta
cat $outdir/${prefix_name}_nonTIR.fasta $outdir/${prefix_name}_TIR.fasta > $outdir/${prefix_name}.fasta
seqkit rmdup -D duplicates -n $outdir/${prefix_name}.fasta > $outdir/${prefix_name}_cds.fasta

##Step C - Translate Nucleotide NBARC sequences - 6-frame translations plus longest ORF frame output, then align and make aa.hmm.
translate.py $outdir/${prefix_name}_cds.fasta $outdir/${prefix_name}_aa.fasta
clustalo -i  $outdir/${prefix_name}_aa.fasta -o  $outdir/${prefix_name}_aa.sto --outfmt=st
hmmbuild -amino $wdir/1.data/${prefix_name}_aa.hmm  $outdir/${prefix_name}_aa.sto

##Step D - Use the E_grandis amino acid NBARC.hmm to search the concatenated euc_NLR_aa.fasta for NBARC domains. We use the -A flag here so that hmmsearch outputs an alignment file. We use easel miniapplications as part of suite in hmmer/3.2 to extract the regions that contain the NBARC domains in a fasta file.
hmmsearch -A $outdir/${final_name}_aa.sto  $wdir/1.data/${prefix_name}_aa.hmm $wdir/1.data/$input_aa.fasta
/usr/local/hmmer/3.2/bin/esl-reformat fasta $outdir/${final_name}_aa.sto > $outdir/${final_name}_aa.fasta
clustalo -t protein -i $outdir/${final_name}_aa.fasta -o $outdir/${final_name}_aa_align.fa --threads=6 --force

## Output from this is used to build phylogeny with iqtree.

module unload seqkit/0.10.1
module unload hmmer/3.3
module unload bedtools/2.29.2
module unload clustal-omega/1.2.4
module unload python/3.6.5
